### DISCUSSION TITLE
How to deploy a self-hosted GPU runner with CUDA enabled

### DISCUSSION THREAD

**Discussion by vaishnavimuley22** [community/community#67167] [Actions]
### Select Topic Area

Question

### Body

Hi, I am using [actions-runner-controller](https://github.com/actions/actions-runner-controller) to deploy a CPU-based and GPU-based self-hosted runner, but facing issues with CUDA in case of GPU runner (like the NVIDIA driver is unable to locate CUDA). 

**Details:**

1. Following is the RunnerDeployment and HorizontalRunnerAutoscaler YAML file:

```YAML
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  name: gpu-runner-deployment
spec:
  template:
    spec:
      organization: MY_ORG_NAME
      containers:
      - name: runner
        volumeMounts:
        - mountPath: /usr/local/cuda
          name: cuda-path
      - name: docker
        volumeMounts:
        - mountPath: /usr/local/cuda
          name: cuda-path
      volumes:
      - name: cuda-path
        hostPath:
          path: /usr/local/cuda
      labels:
        - dev-gpu-runner
      nodeSelector:
        instance-type: gpu
        karpenter.sh/provisioner-name: default-gpu
      tolerations:
      - key: nvidia.com/gpu
        operator: Exists
        effect: NoSchedule
      resources:
        limits:
          cpu: "4.0"
          memory: "32Gi"
          nvidia.com/gpu: 1
        requests:
          cpu: "2.0"
          memory: "16Gi"
          nvidia.com/gpu: 1
---
apiVersion: actions.summerwind.dev/v1alpha1
kind: HorizontalRunnerAutoscaler
metadata:
  name: gpu-runner-deployment-autoscaler
spec:
  scaleDownDelaySecondsAfterScaleOut: 60
  scaleTargetRef:
    name: gpu-runner-deployment
  minReplicas: 0
  maxReplicas: 4
  metrics:
    - type: TotalNumberOfQueuedAndInProgressWorkflowRuns
      repositoryNames:
        - MY_REPO_1
        - MY_REPO_2
        - MY_REPO_3
```

2. As mentioned, there is a karpenter provisioner named **default-gpu** that allocates GPU nodes and the runner pod gets scheduled on it. 

3. Following is the GitHub workflow:
```YAML
name: EKS Runner Testing

on:
    push:
        branches: [vm-gpu-runner-cuda]

jobs:
    build:
        runs-on: dev-gpu-runner

        steps:
        -   name: Checkout code
            uses: actions/checkout@v2

        -   name: Check GPU allocation
            run: |
                nvidia-smi

        -   name: Set up Python 3.x
            uses: actions/setup-python@v4
            with:
                python-version: '3.9'
                architecture: 'x64'

        -   name: Install system dependencies
            run: |
                sudo apt-get update
                sudo apt-get install -y libgl1-mesa-glx ffmpeg libsm6 libxext6 pciutils

        -   name: Check CUDA support
            run: |    
                python -m pip install torch
                python -c "import torch; print(torch.version.cuda); print(torch.cuda.is_available())"

```

4. Here the **nvidia-smi** returns the driver version to be **470.161.03** and CUDA version to be **N/A**, but on setting up the env variables of CUDA within the pod like 
```cmd
echo 'export PATH=/usr/local/cuda/bin:$PATH' >> ~/.bashrc
echo 'export LD_LIBRARY_PATH=/usr/local/cuda/lib64:$LD_LIBRARY_PATH' >> ~/.bashrc
source ~/.bashrc
``` 
the **nvcc -V** command returns the CUDA version to be **11.4**

```cmd
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2021 NVIDIA Corporation
Built on Wed_Jun__2_19:15:15_PDT_2021
Cuda compilation tools, release 11.4, V11.4.48
Build cuda_11.4.r11.4/compiler.30033411_0
```

5. Thus, also added the env variable setup in the github workflow but it didn't work and returned the same CUDA version to be N/A as well as **torch.cuda.is_available()** to be False. 

6. I also tried extending the actions-runner-controllers image and creating a custom self-hosted runner for the same as follows:
```Dockerfile
FROM summerwind/actions-runner:latest

USER root

RUN apt update -y && apt upgrade -y && \
    apt-get install -y python3-pip && \
    apt-get install -y gcc

# Set environment variables for CUDA
ENV LD_LIBRARY_PATH=/usr/local/cuda/lib64:${LD_LIBRARY_PATH}
ENV LD_LIBRARY_PATH /usr/local/nvidia/lib:/usr/local/nvidia/lib64
ENV PATH=/usr/local/cuda/bin:${PATH}

ENV NVIDIA_VISIBLE_DEVICES all
ENV NVIDIA_DRIVER_CAPABILITIES compute,utility
LABEL com.nvidia.volumes.needed="nvidia_driver"

USER runner
```
Built it and pushed it to AWS ECR repo and used it in the workflow as follows:
```YAML
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  name: gpu-runner-deployment
spec:
  template:
    spec:
      organization: MY_ORG_NAME
      dockerdWithinRunnerContainer: true
      image: {acc_id}.dkr.ecr.us-east-1.amazonaws.com/{ecr-repo-name}:self-hosted-custom-runner-image
      labels:
        - dev-gpu-runner
      nodeSelector:
        instance-type: gpu
        karpenter.sh/provisioner-name: default-gpu
      tolerations:
        - key: nvidia.com/gpu
          operator: Exists
          effect: NoSchedule
      resources:
        limits:
          cpu: "4"
          memory: 32Gi
          nvidia.com/gpu: "1"
        requests:
          cpu: "2"
          memory: 16Gi
          nvidia.com/gpu: "1"
---
apiVersion: actions.summerwind.dev/v1alpha1
kind: HorizontalRunnerAutoscaler
metadata:
  name: gpu-runner-deployment-autoscaler
spec:
  scaleDownDelaySecondsAfterScaleOut: 60
  scaleTargetRef:
    name: gpu-runner-deployment
  minReplicas: 0
  maxReplicas: 4
  metrics:
    - type: TotalNumberOfQueuedAndInProgressWorkflowRuns
      repositoryNames:
        - MY_REPO_1
        - MY_REPO_2
        - MY_REPO_3
```

On deploying this and using it in the workflow I started getting the following error message:
```
runnerpod Unregistration started before runner ID is assigned and the runner stopped before obtaining ID within registration timeout. Perhaps the runner successfully ran the job and stopped normally before the runner ID becomes visible via GitHub API? Perhaps the runner pod was terminated by anyone other than ARC? Was it OOM killed? Marking unregistration as completed anyway because there's nothing ARC can do.
```

It would be great if someone help me to figure out what's going wrong and how I can solve it.

**Comment by github-actions** [comment]
🕒 **Discussion Activity Reminder** 🕒

This Discussion has been labeled as dormant by an automated system for having no activity in the last 60 days. Please consider one the following actions:

1️⃣ Close as Out of Date: If the topic is no longer relevant, close the Discussion as `out of date` at the bottom of the page.

2️⃣ Provide More Information: Share additional details or context — or let the community know if you've found a solution on your own.

3️⃣ Mark a Reply as Answer: If your question has been answered by a reply, mark the most helpful reply as the solution.

Note: This dormant notification will only apply to Discussions with the `Question` label. To learn more, see our [recent announcement](https://github.com/orgs/community/discussions/70478).

Thank you for helping bring this Discussion to a resolution! 💬

**Comment by OnkarKDaiv** [comment]
@vaishnavimuley22 I'm also trying to implement the same thing. Just want to check if you got any success in this?
